lib/cove/configuration/contracts/service_contract.rb

```diff
@@ -0,0 +1,45 @@
+require "dry-validation"
+
+module Cove
+  module Configuration
+    module Contracts
+      class ServiceContract < Dry::Validation::Contract
+        params do
+          required(:service_name).filled(:string)
+          required(:image).filled(:string)
+          required(:roles).value(:array, min_size?: 1).each do
+            hash do
+              required(:name).filled(:string)
+              optional(:container_count).filled(:integer)
+              optional(:ingress).value(:array).each do
+                hash do
+                  required(:type).value(included_in?: ["port", "port_range"])
+                  required(:source).value(:filled?)
+                  required(:target).filled(:integer)
+                end
+              end
+            end
+          end
+        end
+
+        rule(:roles).each do
```

Comment:
```suggestion
        rule(:roles).each do |index:|
```

I think you want the index here as well and add it to the key with the error. Or could use `role.name` in the `key` instead.

That reminds me, there should be validation that each role has a unique name.